builtin/providers/aws/resource_aws_vpn_gateway.go

```diff
@@ -0,0 +1,318 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/aws-sdk-go/aws"
+	"github.com/hashicorp/aws-sdk-go/gen/ec2"
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsVpnGateway() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsVpnGatewayCreate,
+		Read:   resourceAwsVpnGatewayRead,
+		Update: resourceAwsVpnGatewayUpdate,
+		Delete: resourceAwsVpnGatewayDelete,
+
+		Schema: map[string]*schema.Schema{
+			"availability_zone": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"vpc_id": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsVpnGatewayCreate(d *schema.ResourceData, meta interface{}) error {
+	ec2conn := meta.(*AWSClient).ec2conn
+
+	createOpts := &ec2.CreateVPNGatewayRequest{
+		AvailabilityZone: aws.String(d.Get("availability_zone").(string)),
+		Type:             aws.String("ipsec.1"),
+	}
+
+	// Create the VPN gateway
+	log.Printf("[DEBUG] Creating VPN gateway")
+	resp, err := ec2conn.CreateVPNGateway(createOpts)
+	if err != nil {
+		return fmt.Errorf("Error creating VPN gateway: %s", err)
+	}
+
+	// Get the ID and store it
+	vpnGateway := resp.VPNGateway
+	d.SetId(*vpnGateway.VPNGatewayID)
+	log.Printf("[INFO] VPN Gateway ID: %s", *vpnGateway.VPNGatewayID)
+
+	// Attach the VPN gateway to the correct VPC
+	return resourceAwsVpnGatewayUpdate(d, meta)
+}
+
+func resourceAwsVpnGatewayRead(d *schema.ResourceData, meta interface{}) error {
+	ec2conn := meta.(*AWSClient).ec2conn
+
+	vpnGatewayRaw, _, err := vpnGatewayStateRefreshFunc(ec2conn, d.Id())()
```

Comment:
Here you call a refresh func `vpnGatewayStateRefreshFunc`, but we only call it once and never try again, like in a `wait.WaitForState` in the `resourceAwsVpnGatewayAttach` function. 

I see that's basically copied from Internet Gateway, so not a fault of yours :smile: . In your experience does this typically take time, and may fail on the first run?